### ISSUE TITLE
[rllib] Custom metrics are not shown if the first episode is not ended while the first training iteration begins

### ISSUE THREAD

**Issue by pengzhenghao** [ray-project/ray#10001]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

Ray 0.8.6
python 3.7
tf 2.1.0

Ray would not report the custom metrics all along if the following condition is met:

* No complete episodes have been collected before the first training iteration starts (so that the first-time metric reporting starts)

However, this is a common cases when we use, say PPO, with large number of workers while the trian batch size is relatively small. For example, we have 10 concurrent environemtns, the episode length of each is 1000, and the train batch size is only 4000 as default. So that the first complete episode is collected at 3rd trainning iteration. If that case happen, the custom metric which records the information of episode would never ever shows up in record.

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

I modified the scritps from `custom_metrics_and_callbacks.py` with 4 points:

1. Change the env to MountainCar so that it's episode length would be much larger than CartPole
2. Set the train batch size and sample batch size to a very low number
3. Set the pole_angles to a random number since we don't use CartPole anymore.
4. Set the terminate iterations from 2000 to 10 for quick exp.

Here is the full script:

```python
"""Example of using RLlib's debug callbacks.

Here we use callbacks to track the average CartPole pole angle magnitude as a
custom metric.
"""

import argparse
import numpy as np

import ray
from ray import tune


def on_episode_start(info):
    episode = info["episode"]
    print("episode {} started".format(episode.episode_id))
    episode.user_data["pole_angles"] = []
    episode.hist_data["pole_angles"] = []


def on_episode_step(info):
    episode = info["episode"]
    # pole_angle = abs(episode.last_observation_for()[2])
    # raw_angle = abs(episode.last_raw_obs_for()[2])
    # assert pole_angle == raw_angle
    pole_angle = np.random.random(1)
    episode.user_data["pole_angles"].append(pole_angle)


def on_episode_end(info):
    episode = info["episode"]
    pole_angle = np.mean(episode.user_data["pole_angles"])
    print("episode {} ended with length {} and pole angles {}".format(
        episode.episode_id, episode.length, pole_angle))
    episode.custom_metrics["pole_angle"] = pole_angle
    episode.hist_data["pole_angles"] = episode.user_data["pole_angles"]


def on_sample_end(info):
    print("returned sample batch of size {}".format(info["samples"].count))


def on_train_result(info):
    print("trainer.train() result: {} -> {} episodes".format(
        info["trainer"], info["result"]["episodes_this_iter"]))
    # you can mutate the result dict to add new fields to return
    info["result"]["callback_ok"] = True


def on_postprocess_traj(info):
    episode = info["episode"]
    batch = info["post_batch"]
    print("postprocessed {} steps".format(batch.count))
    if "num_batches" not in episode.custom_metrics:
        episode.custom_metrics["num_batches"] = 0
    episode.custom_metrics["num_batches"] += 1


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--num-iters", type=int, default=10)
    args = parser.parse_args()

    ray.init()
    trials = tune.run(
        "PG",
        stop={
            "training_iteration": args.num_iters,
        },
        config={
            "env": "MountainCar-v0",
            "train_batch_size": 50,
            "sample_batch_size": 10,
            "callbacks": {
                "on_episode_start": on_episode_start,
                "on_episode_step": on_episode_step,
                "on_episode_end": on_episode_end,
                "on_sample_end": on_sample_end,
                "on_train_result": on_train_result,
                "on_postprocess_traj": on_postprocess_traj,
            },
        },
        return_trials=True)

    # verify custom metrics for integration tests
    custom_metrics = trials[0].last_result["custom_metrics"]
    print(custom_metrics)
    assert "pole_angle_mean" in custom_metrics
    assert "pole_angle_min" in custom_metrics
    assert "pole_angle_max" in custom_metrics
    assert "num_batches_mean" in custom_metrics
    assert "callback_ok" in trials[0].last_result

```

After that, when I check the experiment results, I found that:

![image](https://user-images.githubusercontent.com/22206995/89723567-c95cb880-da2a-11ea-8e6f-e2406b73d1de.png)

You can see that the item `custom_metrics/pole_angle` does not show up, while the assertions of the script are passed. I guess the data is there but the format of `trial_dataframe` is fixed after the first-time reporting, so that the new pop up data would be ignored to save into the trial dataframe?

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by pengzhenghao** [comment]
Note: I know in Ray 0.8.6 we are using an object to represent callbacks, but the issue is still there.
Well, a workaround is add empty term in `on_train_result`:

```python

def on_train_result(info):
    print("trainer.train() result: {} -> {} episodes".format(
        info["trainer"], info["result"]["episodes_this_iter"]))
    # you can mutate the result dict to add new fields to return
    info["result"]["callback_ok"] = True

    if not info["result"]["custom_metrics"]:
        info["result"]["custom_metrics"]["pole_angles_mean"] = np.nan
        # info["result"]["custom_metrics"]["pole_angles_min"] = np.nan
        info["result"]["custom_metrics"]["pole_angles_max"] = np.nan
        info["result"]["custom_metrics"]["num_batches_mean"] = np.nan
        # info["result"]["custom_metrics"]["num_batches_min"] = np.nan
        info["result"]["custom_metrics"]["num_batches_max"] = np.nan

```

The term `custom_metrics/pole_angles_mean` is shown in dataframe while `custom_metrics/pole_angles_min` is not since it is commented.

**Comment by ericl** [comment]
I'm not sure it's possible to provide these semantically before an episode ends for the same reason we can't provide rewards until the episode ends... consider if the final reward is negative. Then, reporting the reward so far (0.0) would be hugely misleading.

**Comment by pengzhenghao** [comment]
I don't mean to predict some metrics that are not collected so far and reporting them. What I think should be added to Ray is the ability to dynamically expand the columns of , at least, the trial_dataframe if new metrics are reported who are not considered by the trial_dataframe at the beginning.

**Comment by pengzhenghao** [comment]
For example, if we set evaluation interval > 1, then in the first iteration the reported stats do not contain anything about evaluation. So that the saved `trial_dataframe` does not have the column for `evaluation/episode_reward_mean` and so on. We can't easily load the data about evaluation from `Analysis` API or from `progress.csv`, even though the data is properly saved at `result.json`
